File: Week1/2 - Happy Number/Solution.py
#
# Created on Fri Apr 10 2020
#
# Title: Leetcode - Happy Number
#
# Author: Vatsal Mistry
# Web: mistryvatsal.github.io
#

class Solution:
    def isHappy(self, n):
        seen = dict()
        while n is not 1:
            summ, curr = 0, n
            while curr is not 0:
                summ += (summ % 10) ** 2
                curr /= 10
            if(seen.__contains__(sum)):
                return False
            seen[summ] = summ
            n = summ
        return True
        

File: Week1/2 - Happy Number/Solution.cpp
/*
 * Created on Mon Apr 27 2020
 *
 * Title: Leetcode - Happy Number
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

#include <iostream>
#include <unordered_set>
#include <cmath>

using namespace std;

class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> seen;
        while(n != 1) {
            int sum = 0;
            int curr = n;
            while(curr != 0) {
                sum += (int)pow((curr % 10), 2);
                curr /= 10;
            }
    
            if(seen.find(sum) != seen.end()) {
                return false;
            }
            seen.insert(sum);
            n = sum;
        }
        return true;
    }
};

int main() {
    int n = 20;
    Solution obj;
    cout << obj.isHappy(n) << endl;
    return 0;
}